python/ray/operator/operator.py

```diff
@@ -0,0 +1,139 @@
+"""
+Ray operator for Kubernetes.
+
+Reads ray cluster config from a k8s ConfigMap, starts a ray head node pod using
+create_or_update_cluster(), then runs an autoscaling loop in the operator pod
+executing this script. Writes autoscaling logs to the directory
+/root/ray-operator-logs.
+
+In this setup, the ray head node does not run an autoscaler. It is important
+NOT to supply an --autoscaling-config argument to head node's ray start command
+in the cluster config when using this operator.
+
+To run, first create a ConfigMap named ray-operator-configmap from a ray
+cluster config. Then apply the manifest at python/ray/autoscaler/kubernetes/operator_configs/operator_config.yaml
+
+For example:
+kubectl create namespace raytest
+kubectl -n raytest create configmap ray-operator-configmap --from-file=python/ray/autoscaler/kubernetes/operator_configs/test_cluster_config.yaml
+kubectl -n raytest apply -f python/ray/autoscaler/kubernetes/operator_configs/operator_config.yaml
+""" # noqa
+import logging
+import multiprocessing as mp
+import os
+import yaml
+
+from ray._private import services
+from ray.autoscaler._private import commands
+from ray import monitor
+from ray.operator import operator_utils
+from ray import ray_constants
+
+
+class RayCluster():
+    def __init__(self, config):
+        self.config = config
+        self.name = self.config["cluster_name"]
+        self.config_path = operator_utils.config_path(self.name)
+
+        self.setup_logging()
+
+        self.subprocess = None
+
+    def do_in_subprocess(self, f, wait_to_finish=False):
+        # First stop the subprocess if it's alive
+        if self.subprocess and self.subprocess.is_alive():
+            self.subprocess.terminate()
+            self.subprocess.join()
+        # Reinstantiate process with f as target and start.
+        self.subprocess = mp.Process(name=self.name, target=f)
```

Comment:
is it possible for us to leak subprocesses if operator.py is killed unexpectedly?

Reply:
I think currently if `operator.py` is killed unexpectedly, the operator pod will shut down.
Which reminds me -- I was going to make all of the ray-clusters managed by the operator fate-share with the operator.
So then everything would go down.
That's of course not optimal behavior.  

Let me also experiment to see what happens if `operator.py` is killed unexpectedly but the pod doesn't go down.

Reply:
I don't think you want to kill all the clusters when operator go down.
It just need to know about them when started.

Reply:
Ah, right. I think that should work with the code as it is.

Reply:
I think Python does an OK  job of cleaning up processes spawned by the multiprocessing module?
To check, I ran a test script that uses multiprocessing to spawn a dummy process that runs forever. After doing a Ctrl-C keyboard interrupt, the pid of the process is no longer present in the output of `ps -ef`.

I'll check that the monitor processes behave in the same way.

Let me know if there's something that can be done to ensure that the processes are cleaned correctly.
(besides rewriting everything to have each cluster's autoscaler `.update` in a for loop, which is probably a good idea to implement sooner or later or sooner to replace this subprocess logic)

Reply:
Your test sounds good enough.

Reply:
Actually, you were right -- terminating `operator.py` did leak the monitor process. (In my test script, the child process was receiving the signal interrupt.)

I've now set the monitor subprocess to be a daemon, and that works -- when I run `kill -SIGTERM <operator_pid>` it stops the monitor subprocess too.
(If you do `kill -SIGKILL <operator_pid>` then the daemon monitor subprocess leaks.)

Reply:
okay then this is still a problem we should fix in the near future.